README.md
<a href="https://www.gotoiot.com/">
    <img src="doc/gotoiot-logo.png" alt="logo" title="Goto IoT" align="right" width="60" height="60" />
</a>

Desarrollo de Aplicaciones Web - Trabajo Final
==============================================

### El trabajo final fue realizado por **Facundo Andrioli Villa**.

Este proyecto es una aplicación web fullstack que se ejecuta sobre el ecosistema `Docker`. Está compuesta por un compilador de `TypeScript` que te permite utilizar este superset de JavaScript para poder programar un `cliente web`. También tiene un servicio en `NodeJS` que te permite ejecutar código en backend y al mismo tiempo disponibilizar el código del cliente web para interactar con el servicio. Además tiene una `base de datos` MySQL que puede interactuar con el backend para guardar y consultar datos, y de manera adicional trae un `administrador` de base de datos para poder administrar la base en caso que lo necesites.

La aplicación IoT de base que viene con este proyecto se encarga de crear una tabla llamada `Devices` en la base de datos, y la idea es que vos puedas desarrollar el código de backend y frontend que te permita controlar desde el navegador el estado de los devices de un hogar inteligente - *como pueden ser luces, TVs, ventiladores, persianas, enchufes y otros* - y almacenar los estados de cada uno en la base de datos. 

Realizando estas tareas vas a a tener una aplicación fullstack IoT del mundo real que utiliza tecnologías actuales en la que un backend es capaz de interactuar con una DB para cumplir con las peticiones de control que se le mandan desde el cliente web.

En esta imagen podés ver una posible implementación del cliente web que controla los artefactos del hogar.

![architecture](doc/webapp_1.png)

## Comenzando 🚀

Esta sección es una guía con los pasos escenciales para que puedas poner en marcha la aplicación.

<details><summary><b>Mira los pasos necesarios</b></summary><br>

### Instalar las dependencias

Para correr este proyecto es necesario que instales `Docker` y `Docker Compose`. 

En [este artículo](https://www.gotoiot.com/pages/articles/docker_installation_linux/) publicado en nuestra web están los detalles para instalar Docker y Docker Compose en una máquina Linux. Si querés instalar ambas herramientas en una Raspberry Pi podés seguir [este artículo](https://www.gotoiot.com/pages/articles/rpi_docker_installation) de nuestra web que te muestra todos los pasos necesarios.

En caso que quieras instalar las herramientas en otra plataforma o tengas algún incoveniente, podes leer la documentación oficial de [Docker](https://docs.docker.com/get-docker/) y también la de [Docker Compose](https://docs.docker.com/compose/install/).

Continua con la descarga del código cuando tengas las dependencias instaladas y funcionando.

### Descargar el código

Para descargar el código, lo más conveniente es que realices un `fork` de este proyecto a tu cuenta personal haciendo click en [este link](https://github.com/Facuu13/DAW-TP-Final). Una vez que ya tengas el fork a tu cuenta, descargalo con este comando (acordate de poner tu usuario en el link):

```
git clone https://github.com/Facuu13/DAW-TP-Final.git
```

> En caso que no tengas una cuenta en Github podes clonar directamente este repo.

### Ejecutar la aplicación

Para ejecutar la aplicación tenes que correr el comando `docker-compose up` desde la raíz del proyecto. Este comando va a descargar las imágenes de Docker de node, de typescript, de la base datos y del admin de la DB, y luego ponerlas en funcionamiento. 

Para acceder al cliente web ingresa a a la URL [http://localhost:8000/](http://localhost:8000/) y para acceder al admin de la DB accedé a [localhost:8001/](http://localhost:8001/). 

Si pudiste acceder al cliente web y al administrador significa que la aplicación se encuentra corriendo bien. 

> Si te aparece un error la primera vez que corres la app, deteńe el proceso y volvé a iniciarla. Esto es debido a que el backend espera que la DB esté creada al iniciar, y en la primera ejecución puede no alcanzar a crearse. A partir de la segunda vez el problema queda solucionado.

</details>

Continuá explorando el proyecto una vez que lo tengas funcionando.

## Configuraciones de funcionamiento 🔩

Al crearse la aplicación se ejecutan los contenedores de Docker de cada servicio, se crea la base de datos y sus tablas. A continuación podés encontrar info si querés cambiar la estructura de la DB o bien sus configuraciones de acceso.

<details><summary><b>Lee cómo configurar la aplicación</b></summary><br>

### Configuración de la DB

Como ya comprobaste, para acceder PHPMyAdmin tenés que ingresar en la URL [localhost:8001/](http://localhost:8001/). En el login del administrador, el usuario para acceder a la db es `root` y contraseña es la variable `MYSQL_ROOT_PASSWORD` del archivo `docker-compose.yml`.

Para el caso del servicio de NodeJS que se comunica con la DB fijate que en el archivo `src/backend/mysql-connector.js` están los datos de acceso para ingresar a la base.

Si quisieras cambiar la contraseña, puertos, hostname u otras configuraciones de la DB deberías primero modificar el servicio de la DB en el archivo `docker-compose.yml` y luego actualizar las configuraciones para acceder desde PHPMyAdmin y el servicio de NodeJS.

### Estructura de la DB

Al iniciar el servicio de la base de datos, si esta no está creada toma el archivo que se encuentra en `db/dumps/smart_home.sql` para crear la base de datos automáticamente.

En ese archivo está la configuración de la tabla `Devices` y otras configuraciones más. Si quisieras cambiar algunas configuraciones deberías modificar este archivo y crear nuevamente la base de datos para que se tomen en cuenta los cambios.

Tené en cuenta que la base de datos se crea con permisos de superusuario por lo que no podrías borrar el directorio con tu usuario de sistema, para eso debés hacerlo con permisos de administrador. En ese caso podés ejecutar el comando `sudo rm -r db/data` para borrar el directorio completo.

</details>


## Detalles principales 🔍

En esta sección vas a encontrar las características más relevantes del proyecto.

<details><summary><b>Mira los detalles más importantes de la aplicación</b></summary><br>
<br>

### Arquitectura de la aplicación

Como ya pudiste ver, la aplicación se ejecuta sobre el ecosistema Docker, y en esta imagen podés ver el diagrama de arquitectura.

![architecture](doc/architecture.png)

### El cliente web

El cliente web es una Single Page Application que se comunica con el servicio en NodeJS mediante JSON a través de requests HTTP. Puede consultar el estado de dispositivos en la base de datos (por medio del servicio en NodeJS) y también cambiar el estado de los mismos. Los estilos del código están basados en **Material Design**.

### El servicio web

El servicio en **NodeJS** posee distintos endpoints para comunicarse con el cliente web mediante requests HTTP enviando **JSON** en cada transacción. Procesando estos requests es capaz de comunicarse con la base de datos para consultar y controlar el estado de los dispositivos, y devolverle una respuesta al cliente web también en formato JSON. Así mismo el servicio es capaz de servir el código del cliente web.

### La base de datos

La base de datos se comunica con el servicio de NodeJS y permite almacenar el estado de los dispositivos en la tabla **Devices**. Ejecuta un motor **MySQL versión 5.7** y permite que la comunicación con sus clientes pueda realizarse usando usuario y contraseña en texto plano. En versiones posteriores es necesario brindar claves de acceso, por este motivo la versión 5.7 es bastante utilizada para fases de desarrollo.

### El administrador de la DB

Para esta aplicación se usa **PHPMyAdmin**, que es un administrador de base de datos web muy utilizado y que podés utilizar en caso que quieras realizar operaciones con la base, como crear tablas, modificar columnas, hacer consultas y otras cosas más.

### El compilador de TypeScript

**TypeScript** es un lenguaje de programación libre y de código abierto desarrollado y mantenido por Microsoft. Es un superconjunto de JavaScript, que esencialmente añade tipos estáticos y objetos basados en clases. Para esta aplicación se usa un compilador de TypeScript basado en una imagen de [Harmish](https://hub.docker.com/r/harmish) en Dockerhub, y está configurado para monitorear en tiempo real los cambios que se realizan sobre el directorio **src/frontend/ts** y automáticamente generar código compilado a JavaScript en el directorio  **src/frontend/js**. Los mensajes del compilador aparecen automáticamente en la terminal al ejecutar el comando **docker-compose up**.

### Ejecución de servicios

Los servicios de la aplicación se ejecutan sobre **contenedores de Docker**, así se pueden desplegar de igual manera en diferentes plataformas. Los detalles sobre cómo funcionan los servicios los podés ver directamente en el archivo **docker-compose.yml**.

### Organización del proyecto

En la siguiente ilustración podés ver cómo está organizado el proyecto para que tengas en claro qué cosas hay en cada lugar.

```sh
├── db                          # directorio de la DB
│   ├── data                    # estructura y datos de la DB
│   └── dumps                   # directorio de estructuras de la DB
│       └── smart_home.sql      # estructura con la base de datos "smart_home"
├── doc                         # documentacion general del proyecto
└── src                         # directorio codigo fuente
│   ├── backend                 # directorio para el backend de la aplicacion
│   │   ├── index.js            # codigo principal del backend
│   │   ├── mysql-connector.js  # codigo de conexion a la base de datos
│   │   ├── package.json        # configuracion de proyecto NodeJS
│   │   └── package-lock.json   # configuracion de proyecto NodeJS
│   └── frontend                # directorio para el frontend de la aplicacion
│       ├── js                  # codigo javascript que se compila automáticamente
│       ├── static              # donde alojan archivos de estilos, imagenes, fuentes, etc.
│       ├── ts                  # donde se encuentra el codigo TypeScript a desarrollar
│       └── index.html          # archivo principal del cliente HTML
├── docker-compose.yml          # archivo donde se aloja la configuracion completa
├── README.md                   # este archivo
├── CHANGELOG.md                # archivo para guardar los cambios del proyecto
├── LICENSE.md                  # licencia del proyecto
```

> No olvides ir poniendo tus cambios en el archivo `CHANGELOG.md` a medida que avanzas en el proyecto.

</details>

## Detalles de implementación 💻

En esta sección podés ver los detalles específicos de funcionamiento del código y que son los siguientes.

<details><summary><b>Mira los detalles de implementación</b></summary><br>

### Agregar un dispositivo

1. En la siguiente imagen se puede observar la vista del tablero, hacemos clic en **Agregar Dispositivo**
![architecture](doc/agregar_1.png)
2. Luego se va abrir un popup en el que se va a poder ingresar los datos del nuevo dispositivo:
    - Nombre del dispositivo.
    - Descripcion del dispositivo.
    - Estado (Encendido o Apagado).
    - Si queremos manejar la intensidad del dispositivo (Si o No). 
    - Intensidad (numero que va de 0 a 1, con pasos de 0.1)
![architecture](doc/agregar_2.png)
3. El proximo paso es llenar los datos del nuevo dispositivo y hacer clic en **Guardar**, en caso de no querer guardar el nuevo dispositivo apretamos **Cancelar**
![architecture](doc/agregar_3.png)
4. Como se ve en la siguiente imagen, el dispositivo agregado se puede observar en el tablero.
![architecture](doc/agregar_4.png)


### Frontend

El tablero de dispositivos se ve de la siguiente manera:
![architecture](doc/webapp_1.png)
- En la imagen anterior se puede observar una lista de los dispositivos que estan disponibles.
- Se puede observar al principio de la pagina un boton llamado **Agregar Dispositivo**, el cual es un modal, por ende se va abrir un **popup** y vamos a poder ingresar los datos del nuevo dispositivo. El funcionamiento de este boton se explico en el punto anterior.
- Cuando se inicia la pagina se llama a una funcion **showDevices** que nos va a permitir visualizar los dispositivos que estan agregados en la base de datos. Estas tarjetitas de devices se van a generar dinamicamente.
- El div de este tablaro va estar definido por **"col s12 m6 lg3"** por lo que:
    - En pantallas pequeñas veremos un dispositivo.
    - En pantallas medianas veremos dos dispositivos.
    - En pantallas grandes veremos cuatros dispositivos.
- Estas tarjetas de devices estan conformado por lo siguiente:
    - Nombre del dispositivo.
    - Descripcion del dispositivo.
    - Estado.
    - Un range que nos permite manejar la intensidad del dispositivo.
    - Un boton **Eliminar** para borrar el dispositivo.
    - Un boton **Editar** para editar los datos del dispositivo.

        ![architecture](doc/device_1.png)

- Dependiendo del estado del dispositivo el icono en forma de rayito va ir cambiando como se puede observar en la siguiente imagen:
    
    ![architecture](doc/device_2.png)

- Cuando se haga clic en el boton **Eliminar** saldra un cartel de confirmacion para asegurar que estamos seguro de eliminarlo.
    ![architecture](doc/eliminar_1.png)

- El boton **Editar** es un modal, cuando se hace clic se abrirá un **popup** y nos mostrara los valores que tiene el dispositivo actualmente y podremos editarlos:
    ![architecture](doc/editar_1.png)

- Por ultimo, cuando un dispositivo tenga disponible para manejar la intensidad si esta encendido el range va estar disponible, pero si este se encuentra apagado el range va a quedar deshabilitado.

### Backend

#### Endpoints

- GET:
    - app.get('/devices/'): Nos va a devolver la lista de todos los dispositivos.
    - app.get('/devices/:id'): Nos va a devolver el dispositivo que le indiquemos.

- DELETE:
    - app.delete('/devices/:id'): Se eliminara el dispositivo que le indiquiemos.

- POST:
    - app.post('/device'): metodo para agregar un nuevo dispositivo.
    - app.post('/deviceState'): metodo para actualizar el state de un dispositivo.
    - app.post('/deviceIntensity'): metodo para actualizar el intensity de un dispositivo.

- PUT:
    - app.put('/devices/:id'): metodo para actualizar los datos de un dispositivo.

#### Funciones

Funciones que interacionan con el backend y el cliente web:

- **private showDevices()**: funcion que sirve para mostrar los dispositivos en la pagina y para hacer un refresh.
- **private ejecutarPost(device:Device)**: funcion para ejecutar el metodo POST para agregar un nuevo dispositivo.
- **private ejecutarPostState(id:number,state:boolean)**: funcion para ejecutar el metodo POST para actualizar el state de los dispositivo.
- **private ejecutarPostIntensity(id:number,intensity:number)**: funcion para ejecutar el metodo POST para actualizar el intensity de los dispositivo.
- **private ejecutarDelete(id:string)**: funcion para ejecutar el metodo DELETE para eliminar un dispositivo
- **private ejecutarPUT(device:Device,id:number)**: funcion para ejecutar el metodo PUT para actualizar los datos de un dispositivo

#### Base de datos

Se agrego a la tabla **Devices** un nuevo campo llamado **intensity** para poder manejar la intensidad de los dispositivos

La tabla **Devices** quedo conformada de la siguiente manera:
- id
- name
- description
- state
- type
- intensity

</details>

</details>


## Tecnologías utilizadas 🛠️

En esta sección podés ver las tecnologías más importantes utilizadas.

<details><summary><b>Mira la lista completa de tecnologías</b></summary><br>

* [Docker](https://www.docker.com/) - Ecosistema que permite la ejecución de contenedores de software.
* [Docker Compose](https://docs.docker.com/compose/) - Herramienta que permite administrar múltiples contenedores de Docker.
* [Node JS](https://nodejs.org/es/) - Motor de ejecución de código JavaScript en backend.
* [MySQL](https://www.mysql.com/) - Base de datos para consultar y almacenar datos.
* [PHPMyAdmin](https://www.phpmyadmin.net/) - Administrador web de base de datos.
* [Material Design](https://material.io/design) - Bibliotecas de estilo responsive para aplicaciones web.
* [TypeScript](https://www.typescriptlang.org/) - Superset de JavaScript tipado y con clases.

</details>

## Contribuir 🖇️

Si estás interesado en el proyecto y te gustaría sumar fuerzas para que siga creciendo y mejorando, podés abrir un hilo de discusión para charlar tus propuestas en [este link](https://github.com/gotoiot/app-fullstack-base/issues/new). Así mismo podés leer el archivo [Contribuir.md](https://github.com/gotoiot/gotoiot-doc/wiki/Contribuir) de nuestra Wiki donde están bien explicados los pasos para que puedas enviarnos pull requests.

## Sobre Goto IoT 📖

Goto IoT es una plataforma que publica material y proyectos de código abierto bien documentados junto a una comunidad libre que colabora y promueve el conocimiento sobre IoT entre sus miembros. Acá podés ver los links más importantes:

* **[Sitio web](https://www.gotoiot.com/):** Donde se publican los artículos y proyectos sobre IoT. 
* **[Github de Goto IoT:](https://github.com/gotoiot)** Donde están alojados los proyectos para descargar y utilizar. 
* **[Comunidad de Goto IoT:](https://groups.google.com/g/gotoiot)** Donde los miembros de la comunidad intercambian información e ideas, realizan consultas, solucionan problemas y comparten novedades.
* **[Twitter de Goto IoT:](https://twitter.com/gotoiot)** Donde se publican las novedades del sitio y temas relacionados con IoT.
* **[Wiki de Goto IoT:](https://github.com/gotoiot/doc/wiki)** Donde hay información de desarrollo complementaria para ampliar el contexto.

## Muestas de agradecimiento 🎁

Si te gustó este proyecto y quisieras apoyarlo, cualquiera de estas acciones estaría más que bien para nosotros:

* Apoyar este proyecto con una ⭐ en Github para llegar a más personas.
* Sumarte a [nuestra comunidad](https://groups.google.com/g/gotoiot) abierta y dejar un feedback sobre qué te pareció el proyecto.
* [Seguirnos en twitter](https://github.com/gotoiot/doc/wiki) y dejar algún comentario o like.
* Compartir este proyecto con otras personas.

## Autores 👥

Las colaboraciones principales fueron realizadas por:

* **[Agustin Bassi](https://github.com/agustinBassi)**: Ideación, puesta en marcha y mantenimiento del proyecto.
* **[Ernesto Giggliotti](https://github.com/ernesto-g)**: Creación inicial del frontend, elección de Material Design.

CHANGELOG.md
# DAW Base App - Changes Log

Todos los cambios notables en este proyecto se documentarán en este archivo.

El formato se basa en [Keep a Changelog](http://keepachangelog.com/) y este proyecto se adhiere al [Semantic Versioning](http://semver.org/).

## 1.2
* Documentacion:
    * Se actualizo la documentacion en el archivo Readme.md

## 1.1

* Base de datos:
    * Se agrego a la tabla **Devices** un nuevo campo llamado **intensity** para poder manejar la intensidad de los dispositivos
* Backend:
    * Se actualizaron los metodos para que se pueda manejar esta nueva caracteristica.
* Frontend:
    * Se hicieron las modificaciones pertinentes para que se pueda manejar la intensidad, esto se va a poder manejar desde el range.

## 1.0

* Backend:
    * Funcionalidad para borrar un dispositivo.
    * Funcionalidad para agregar un nuevo dispositivo.
    * Tener la posibilidad desde el backend de no solo actualizar el valor del estado, sino de actualizar cualquier valor.
* Frontend:
    * Edición de dispositivos existentes. Nombre, descripción, tipo.
    * Posibilidad de agregar o quitar dispositivos.
    * Posibilidad de indicar qué valores de estado puede asumir el dispositivo.
    * También se implemento un range pero no tiene ningun efecto, solo estetico.

src/frontend/ts/main.ts
var M;
class Main implements EventListenerObject{
    //inicializamos la variable global device
    device:Device = new Device();

    constructor() {
        this.showDevices(); //Llamamos la funcion de mostrar dispositivos para 
        //que cuando inicie la pagina, se muestren de manera automatica
    }

    //Funcion que sirve para mostrar los dispositivos y para hacer un refresh
    private showDevices(){
        let xmlRequest = new XMLHttpRequest();
        xmlRequest.onreadystatechange = ()=> {
            if(xmlRequest.readyState == 4){
                if(xmlRequest.status == 200){
                    let respusta = xmlRequest.responseText;
                    let datos: Array<Device> = JSON.parse(respusta);
                    let div = document.getElementById("listaDisp");
                    div.innerHTML = '';
                    for (let d of datos){
                        // Asigna 'checked' a isChecked si d.state es verdadero, de lo contrario, asigna una cadena vacía.
                        const isChecked = d.state ? 'checked' : '';
                        // Asigna 'disable' a isChecked si d.state es falso, de lo contrario, asigna una cadena vacía.
                        //esto se hace para habilitar el range
                        const isDisabled = d.state ? '' : 'disabled';
                        // Generamos un ID único para el dispositivo
                        const deviceId = `${d.id}`;
                        let type: string;
                        
                        let deviceDiv = document.createElement("div"); 
                        deviceDiv.className = "col s12 m6 lg3";
                        deviceDiv.id = deviceId;

                        // Generamos un ID único para el checkbox basado en el ID del dispositivo
                        const checkboxId = `cb_${d.id}`;
                        // Generamos un ID único para el range basado en el ID del dispositivo
                        
                        const rangeId = `rg_${d.id}`;
                        //dependiendo del estado, se va ir alternando el icono de encendido o apagado
                        const icon= d.state ? 'flash_on' : 'flash_off';;

                        type = `
                        <div class="switch">
                            <label class=label-checkbox>
                                Off
                                <input type="checkbox" id="${checkboxId}" ${isChecked}>
                                <span class="lever"></span>
                                On
                            </label>
                        </div>`
                        if(d.type == 1){
                            
                            type += `<form action="#">
                            <p class="range-field">
                                <input type="range" id="${rangeId}" min="0" max="1" step="0.1" value="${d.intensity}" ${isDisabled}/>
                            </p>
                            </form>`
                        }

                        deviceDiv.innerHTML += `<div class=device-container><i class="small material-icons">${icon}</i>
                        <h4>${d.name}</h4>
                        <p>${d.description}</p>
                        ${type}
                        </div>`
                        div.appendChild(deviceDiv);
                        this.crearBotones(deviceDiv,div,d);

                    }

                    for(let d of datos){
                        
                        let checkbox = document.getElementById(`cb_${d.id}`);
                        checkbox.addEventListener("click",this);
                        
                        if(d.type == 1){
                            let range = document.getElementById(`rg_${d.id}`);
                            range.addEventListener("change",this);
                        }
                    }
                }
                    
            }
        
        }
        xmlRequest.open("GET","http://localhost:8000/devices",true); //lo ponemos en true para que se ejecute de forma asincrona
        xmlRequest.send();
    }

    //Funcion para editar los valores de un dispositivos
    private editarDevice(device:Device,name_id:string,desc_id:string,state_id:string,type_id:string){
        // Obtener el elemento input por su ID
        const nombre = document.getElementById(`${name_id}`) as HTMLInputElement;
        const description = document.getElementById(`${desc_id}`) as HTMLInputElement;
        // Obtener el elemento select por su ID
        const state = document.getElementById(`${state_id}`) as HTMLSelectElement;
        const type = document.getElementById(`${type_id}`) as HTMLSelectElement;

        //Obtenemos los valores
        const n = nombre.value;
        const d = description.value;
        const s = Number(state.value);
        const t = Number(type.value);

        //actualizamos los valores
        this.device.name = n;
        this.device.description = d;
        this.device.state = s;
        this.device.type =t;


        this.ejecutarPUT(this.device,device.id); //llamamos al metodo put para actualizar los valores
        this.showDevices();//refresh   
    }

    //Funcion para eliminar un dispositivo
    private eliminarDevice(id:string){
        // Mostrar un mensaje de confirmación
        const confirmDelete = window.confirm("¿Estás seguro de eliminar este dispositivo?");
        if (confirmDelete) {
            const miDiv= document.getElementById(id);
            if(miDiv){
                console.log("eliminamos device",id)
                this.ejecutarDelete(id);
                this.showDevices(); //refresh

            }
        }
        else{
            //no hacemos nada
        }

    }

    //Funcion para crear los botones de los dispositivos
    private crearBotones(deviceDiv,div,d){
        
        //creamos una variable para la id del modal
        let modal_id: string;
        modal_id = `modal${deviceDiv.id}`;

        // Creamos el boton para editar como modal
        const a = document.createElement("a");
        a.className = "waves-effect waves-light btn modal-trigger transparent";
        a.href = `#${modal_id}`;
        a.textContent = "Editar";

        //Id unica para cada boton de editar
        let btn_edit_id : string;
        btn_edit_id = `btnEditar_${deviceDiv.id}`;

        //id unicas para el modal
        let name_id : string;
        name_id = `eNombre${deviceDiv.id}`;

        let desc_id : string;
        desc_id = `eDescription${deviceDiv.id}`;

        let state_id : string;
        state_id = `eState${deviceDiv.id}`;

        let type_id : string;
        type_id = `eType${deviceDiv.id}`;

        let estado_device = d.state ? 'Encendido' : 'Apagado';

        let manejar_intensidad = d.type ? 'Si' : 'No';
        // se generala el html del modal de manera dinamica para cada device
        const htmlModal = `
        <div id="${modal_id}" class="modal" style="display: none;">
            <div class="modal-content">
                <h4>Editar dispositivo</h4>
                    <div>
                        <label for="${name_id}">Nombre del dispositivo</label>
                        <input id="${name_id}" type="text" value="${d.name}" />
                    </div>
                    <div >
                        <label for="${desc_id}">Descripción del dispositivo</label>
                        <input id="${desc_id}" type="text"  value="${d.description}" />
                    </div>
                    <div>
                    <label>Estado del dispositivo</label>
                        <select id="${state_id}">
                            <option value="" disable selected>${estado_device}</option>
                            <option value="0">Apagado</option>
                            <option value="1">Encendido</option>
                        </select>
                    </div>
                    <div>
                    <label>Manejar intensidad?</label>
                        <select id="${type_id}">
                            <option value="" disable selected>${manejar_intensidad}</option>
                            <option value="0">No</option>
                            <option value="1">Si</option>
                        </select>
                    </div>
            </div>
            <div class="modal-footer">
                <a href="#!" class="modal-close waves-effect waves-green btn-flat">Cancelar</a>
                <button class="modal-close waves-effect waves-green btn-flat" id="${btn_edit_id}">Guardar</button>
            </div>
        </div>
        `;

        // Creamos el boton para eliminar
        const buttonEliminar = document.createElement("button");
        buttonEliminar.className="waves-effect waves-light btn modal-trigger transparent";
        buttonEliminar.textContent = "Eliminar";

        //cuando hacemos clic llama a la funcion
        buttonEliminar.onclick = () => {
        this.eliminarDevice(deviceDiv.id);
        };

        deviceDiv.appendChild(a);  //Agregamos el modal
        deviceDiv.innerHTML += htmlModal;

        // Inicializar el modal dinámico después de agregarlo al DOM
        const modalElement = deviceDiv.querySelector(`#${modal_id}`);
        M.Modal.init(modalElement, null);

        // Inicializar los selects dentro del modal
        const selectElements = deviceDiv.querySelectorAll('select');
        M.FormSelect.init(selectElements, null);
        
        deviceDiv.appendChild(buttonEliminar);//agrega el boton Eliminar al final del div

        div.appendChild(deviceDiv);

        //Asignamos la funcion al boton guardar de editar
        let botonEditar = document.getElementById(btn_edit_id)
        botonEditar.addEventListener("click",()=>{
            this.editarDevice(d,name_id,desc_id,state_id,type_id);
        })
    }

    //Funcion para agregar nuevos dispositivos
    private agregarDevice(){
        // Obtener el elemento input por su ID
        const nombre = document.getElementById("iNombre") as HTMLInputElement;
        const description = document.getElementById("iDescription") as HTMLInputElement;
        // Obtener el elemento select por su ID
        const state = document.getElementById("iState") as HTMLSelectElement;
        const type = document.getElementById("iType") as HTMLSelectElement;

        // Obtener el elemento input por su ID
        const intensity = document.getElementById("iIntensidad") as HTMLInputElement;
  
        //Obtenemos los valores del modal
        const n = nombre.value;
        const d = description.value;
        const s = Number(state.value);
        const t = Number(type.value);
        const i = Number(intensity.value);

        this.device.name = n;
        this.device.description = d;
        this.device.state = s;
        this.device.type =t;
        this.device.intensity = i;

        this.ejecutarPost(this.device); //llamamos al metodo post para agregar el nuevo dispositivo
        this.showDevices();//refresh
    }

    //Funcion para ejecutar el metodo POST para agregar un nuevo dispositivo
    private ejecutarPost(device:Device){
        let xmlRequest = new XMLHttpRequest();
        xmlRequest.onreadystatechange = ()=> {
            if(xmlRequest.readyState == 4){
                if(xmlRequest.status == 200){
                    console.log("llego respuesta",xmlRequest.responseText);
                }
            }

        }
        xmlRequest.open("POST","http://localhost:8000/device",true); //lo ponemos en true para que se ejecute de forma asincrona
        xmlRequest.setRequestHeader("Content-Type","application/json"); //se indica el formato en el que se va enviar la informacion
        xmlRequest.send(JSON.stringify(device));
    }

    //Funcion para ejecutar el metodo POST para actualizar el state de los dispositivo
    private ejecutarPostState(id:number,state:boolean){
        let xmlRequest = new XMLHttpRequest();
        xmlRequest.onreadystatechange = ()=> {
            if(xmlRequest.readyState == 4){
                if(xmlRequest.status == 200){
                    console.log("llego respuesta",xmlRequest.responseText);
                }
            }

        }
        xmlRequest.open("POST","http://localhost:8000/deviceState",true); //lo ponemos en true para que se ejecute de forma asincrona
        xmlRequest.setRequestHeader("Content-Type","application/json"); //se indica el formato en el que se va enviar la informacion
        let s = {id:id,
                state:state};
        xmlRequest.send(JSON.stringify(s));
    }

    //Funcion para ejecutar el metodo POST para actualizar el intensity de los dispositivo
    private ejecutarPostIntensity(id:number,intensity:number){
        let xmlRequest = new XMLHttpRequest();
        xmlRequest.onreadystatechange = ()=> {
            if(xmlRequest.readyState == 4){
                if(xmlRequest.status == 200){
                    console.log("llego respuesta",xmlRequest.responseText);
                }
            }

        }
        xmlRequest.open("POST","http://localhost:8000/deviceIntensity",true); //lo ponemos en true para que se ejecute de forma asincrona
        xmlRequest.setRequestHeader("Content-Type","application/json"); //se indica el formato en el que se va enviar la informacion
        let s = {id:id,
                intensity:intensity};
        xmlRequest.send(JSON.stringify(s));
    }

    //Funcion para ejecutar el metodo DELETE
    private ejecutarDelete(id:string){
        let xmlRequest = new XMLHttpRequest();
        xmlRequest.onreadystatechange = ()=> {
            if(xmlRequest.readyState == 4){
                if(xmlRequest.status == 200){
                }
            }
        }
        const url = `http://localhost:8000/devices/${id}`; // incorporamos el ID en la URL
        xmlRequest.open("DELETE",url,true);
        xmlRequest.send(null);
    }

    //Funcion para ejecutar el metodo PUT
    private ejecutarPUT(device:Device,id:number){
        let xmlRequest = new XMLHttpRequest();
        xmlRequest.onreadystatechange = ()=> {
            if(xmlRequest.readyState == 4){
                if(xmlRequest.status == 200){
                    console.log("llego put",xmlRequest.responseText);
                }
            }
        }
        const url = `http://localhost:8000/devices/${id}`; // incorporamos el ID en la URL
        xmlRequest.open("PUT",url,true);
        xmlRequest.setRequestHeader("Content-Type","application/json"); //se indica el formato en el que se va enviar la informacion
        xmlRequest.send(JSON.stringify(device));
    }



    handleEvent(object: Event): void {
        let elemento = <HTMLElement> object.target;
        console.log(elemento.id)
        if("btnAgregar" === elemento.id){
            this.agregarDevice();
        }else if(elemento.id.startsWith("cb_")){ //elemento.checked para saber si esta en true o false
            let checkbox = <HTMLInputElement>elemento; //casteamos
            console.log(checkbox.checked,elemento.id.substring(3,4))
            this.ejecutarPostState(Number(elemento.id.substring(3,elemento.id.length)),checkbox.checked);
            this.showDevices(); //refresh
        }else if(elemento.id.startsWith("rg_")){
            let range = <HTMLInputElement>elemento; //casteamos
            console.log(range.value);
            this.ejecutarPostIntensity(Number(elemento.id.substring(3,elemento.id.length)),Number(range.value))
            this.showDevices(); //refresh
        }
    }

}
window.addEventListener("load",  ()=> {

    var elemsModal = document.querySelectorAll('.modal');
    M.Modal.init(elemsModal, null);

    var elems = document.querySelectorAll('select');
    M.FormSelect.init(elems, null);

    // Creamos una instancia de la clase Main
    let main: Main = new Main();

    let botonAgregar = document.getElementById("btnAgregar")
    botonAgregar.addEventListener("click",main)

});